### ISSUE TITLE
Feature request: copy as HTML

### ISSUE THREAD

**Issue by 0xabu** [mintty/mintty#811]
mintty supports copying the selection as rich text format, however I've noticed that some "modern" Windows apps don't seem to support rich text and fall back to pasting plain text -- this includes the Mail app, and the OneNote app, for example. A workaround I've found for this is to copy from mintty, paste into Word, which imports/renders the rich text, then paste into the target application. I'm guessing (but haven't verified) that Word is converting to HTML here; if so, it would be nice if mintty natively supported copying the HTML.

**Comment by mintty** [comment]
Mintty supports Export as HTML already (HTML Screen Dump from extended context menu), it's copied to a file, however.
To copy HTML to the clipboard, a suitable format needs to be selected. HTML is not among the [Standard Clipboard Formats](https://docs.microsoft.com/de-de/windows/desktop/dataxchg/standard-clipboard-formats), but there is also an [HTML Clipboard Format](https://docs.microsoft.com/de-de/windows/desktop/dataxchg/html-clipboard-format); why it's separately described is a bit obscure however.
Before I even consider trying that, I'd need an application that would in fact paste HTML, for testing.

**Comment by 0xabu** [comment]
Either of the built-in Windows 10 apps I mentioned (Mail, OneNote) would be good candidates.

I just did some tests with the debugger to try to dig into how this works. mintty calls user32!SetClipboardData with three formats: 0xd (Unicode text), 1 (plain text), and 0xc09f (RTF). Word retrieves the RTF format data on paste from mintty. When I copy from Word, it calls SetClipboardData for a whole plethora of formats:

1. 0xc009 DataObject
1. 0xc00e Object Descriptor
1. 0xc09f Rich Text Format
1. 0xc0bb HTML Format
1. 1 Plain text
1. 0xd Unicode
1. 0xe Enhanced metafile handle
1. 3 Metafile picture
1. 0xc00b Embed Source
1. 0xc004 Native
1. 0xc003 OwnerLink
1. 0xc00d Link Source
1. 0xc00f Link Source Descriptor
1. 0xc002 ObjectLink
1. 0xc013 Ole Private Data

However, in the target apps, I don't see any hits on GetClipboardData -- it seems the clipboard API in UWP is entirely different. So I can't confirm they are using HTML, but it seems the most likely explanation.

**Comment by mintty** [comment]
> 4. 0xc0bb HTML Format

Where do you have this definition from? I see no defined value for CF_HTML in the docs or in the Windows SDK (WinUser.h does not mention CF_HTML at all). Also given that the description of the HTML Clipboard Format is utterly obscure and examples are broken, I don't see a promising way to implement this.

**Comment by 0xabu** [comment]
That's just the string returned by GetClipboardFormatNameA(0xc0bb), which I called inside the debugger, to figure out what formats Word was emitting. I don't know if that value is global or unique to my system -- from a quick read of the docs it looks like you are supposed to call something like [RegisterClipboardFormat](https://docs.microsoft.com/en-us/windows/desktop/api/Winuser/nf-winuser-registerclipboardformata)("HTML Format"), and it returns the ID.

Here's an example that came up in a quick search:
https://support.microsoft.com/en-us/help/274308/how-to-add-html-code-to-the-clipboard-by-using-visual-c

**Comment by 0xabu** [comment]
Another old example: https://blogs.msdn.microsoft.com/jmstall/2007/01/21/copying-html-on-the-clipboard/

It looks like the format is quite complex, but if we're lucky just pasting a prepared header on the front of standard HTML may be good enough.

**Comment by 0xabu** [comment]
What I wrote above about pasting a prepared header is bogus -- there are byte offsets in the header, but if you don't have context it's not terribly complex. In any case, this is the best summary of the format, sample code, and pointers to other references on the topic that I found: https://theartofdev.com/2014/06/12/setting-htmltext-to-clipboard-revisited/

**Comment by mintty** [comment]
Anyway, from these hints I guess I can construct something. A browser apparently copies this format, too, so it can be read to check the format. For the desired direction however, copying to the clipboard, I need an application that can paste it, as requested, for testing. Not something speculative but something I can use, please.

**Comment by 0xabu** [comment]
I think the problem with that example is that something has appended extra trailing whitepace in the version on the website. If I trim the trailing whitespace, at least the "StartHTML" field makes sense:
```
$ cat test.txt
Version:0.9
StartHTML:000000149
EndHTML:000000329
StartFragment:000000266
EndFragment:000000298
StartSelection:000000266
EndSelection:000000298%
$ wc -c test.txt
149 test.txt
```
I believe the EndHTML is also correct if you account for mangling of the UTF8.

Do you have access to Windows 10? As I said, the Mail and OneNote apps that come with it both paste this format. I'm checking whether there's a good (non-UWP) app that you can use.

**Comment by mintty** [comment]
> StartHTML

Yes, but StartFragment and StartSelection make no sense at all. Also there is no clear explanation about these two things which are the same in all examples.

**Comment by 0xabu** [comment]
Here's an open source Win32 app that supports HTML paste: http://openlivewriter.org/

https://github.com/OpenLiveWriter/OpenLiveWriter/blob/master/src/managed/OpenLiveWriter.CoreServices/DataObject/HTMLDataObject.cs seems to be the main parser.

The downside is that you need to create a throwaway account on a blog site (if you don't already have one) to use it.

**Comment by 0xabu** [comment]
Curiously they emit using header version 1.0 that [Raymond Chen's blog ](http://blogs.msdn.com/b/oldnewthing/archive/2013/01/18/10386079.aspx) explicitly says is broken :)

**Comment by mintty** [comment]
Test case: copy some HTML from a browser (Firefox):
I can read clipboard format "text/html" from the clipboard as wide characters (UTF-16), but the contents is plain HTML, no preceding description tags.
I can also read clipboard format "HTML Format" as single-byte characters (actual encoding to be checked), and I get an initial line "Version:0.9" but no further tags, then the plain text, no HTML tags.
Hmm.

**Comment by mintty** [comment]
Released 2.9.4; note that HTML does not include Sixel graphics (yet).

**Comment by mintty** [comment]
Released 2.9.5. HTML copy not enabled by default, but added to Options menu.
Also various HTML copy options available in extended context menu (Ctrl+right-click).

**Comment by 0xabu** [comment]
> Thanks! Sorry I missed this when you first made the commit. I've tested and have some compat-related improvements to the HTML output. I'll clean them up and send you a PR to review.

I finally got around to doing this, but it turns out the fixes you already made for "tools like PowerPoint" in c2f48e0b00111e2dc49b2d43dee799d703bed7b1 also subsume my changes. Thanks again!
